File: typescript/src/did.ts
// SPDX-License-Identifier: BSL-1.1
// Copyright (c) 2026 MuVeraAI Corporation

import type { DIDDocument, DIDMethod, VerificationMethod } from "./types.js";

/**
 * Multicodec prefix for Ed25519 public keys: 0xed 0x01.
 * Used in did:key encoding per the multicodec specification.
 */
const ED25519_MULTICODEC_PREFIX = new Uint8Array([0xed, 0x01]);

/**
 * Base58btc alphabet (Bitcoin alphabet, used by did:key).
 */
const BASE58_ALPHABET =
  "123456789ABCDEFGHJKLMNPQRSTUVWXYZabcdefghijkmnopqrstuvwxyz";

/**
 * Encodes a Uint8Array to base58btc (Bitcoin encoding).
 */
export function encodeBase58Btc(bytes: Uint8Array): string {
  const digits: number[] = [0];
  for (const byte of bytes) {
    let carry = byte;
    for (let i = 0; i < digits.length; i++) {
      carry += (digits[i] ?? 0) << 8;
      digits[i] = carry % 58;
      carry = Math.floor(carry / 58);
    }
    while (carry > 0) {
      digits.push(carry % 58);
      carry = Math.floor(carry / 58);
    }
  }

  let result = "";
  for (const byte of bytes) {
    if (byte !== 0) break;
    result += "1";
  }
  for (let i = digits.length - 1; i >= 0; i--) {
    result += BASE58_ALPHABET[digits[i] ?? 0];
  }
  return result;
}

/**
 * Decodes a base58btc string to Uint8Array.
 */
export function decodeBase58Btc(encoded: string): Uint8Array {
  const bytes: number[] = [0];
  for (const char of encoded) {
    const value = BASE58_ALPHABET.indexOf(char);
    if (value < 0) throw new Error(`Invalid base58btc character: ${char}`);
    let carry = value;
    for (let i = 0; i < bytes.length; i++) {
      carry += (bytes[i] ?? 0) * 58;
      bytes[i] = carry & 0xff;
      carry >>= 8;
    }
    while (carry > 0) {
      bytes.push(carry & 0xff);
      carry >>= 8;
    }
  }

  let leadingZeros = 0;
  for (const char of encoded) {
    if (char !== "1") break;
    leadingZeros++;
  }

  const result = new Uint8Array(leadingZeros + bytes.length);
  for (let i = 0; i < bytes.length; i++) {
    result[leadingZeros + i] = bytes[bytes.length - 1 - i] ?? 0;
  }
  return result;
}

/**
 * Derives a did:key DID from a raw Ed25519 public key (32 bytes).
 * Encoding: multibase base58btc prefix 'z' + base58btc(0xed01 || publicKeyBytes).
 */
export function deriveKeyDID(publicKey: Uint8Array): string {
  if (publicKey.length !== 32) {
    throw new Error(
      `deriveKeyDID: expected 32-byte Ed25519 public key, got ${publicKey.length}`,
    );
  }
  const prefixed = new Uint8Array(
    ED25519_MULTICODEC_PREFIX.length + publicKey.length,
  );
  prefixed.set(ED25519_MULTICODEC_PREFIX);
  prefixed.set(publicKey, ED25519_MULTICODEC_PREFIX.length);
  return "did:key:z" + encodeBase58Btc(prefixed);
}

/**
 * Extracts the raw Ed25519 public key bytes from a did:key DID.
 */
export function extractPublicKeyFromKeyDID(did: string): Uint8Array {
  if (!did.startsWith("did:key:z")) {
    throw new Error(`extractPublicKeyFromKeyDID: not a base58btc did:key: ${did}`);
  }
  const encoded = did.slice("did:key:z".length);
  const decoded = decodeBase58Btc(encoded);

  if (decoded.length < ED25519_MULTICODEC_PREFIX.length) {
    throw new Error("extractPublicKeyFromKeyDID: decoded bytes too short");
  }
  if (
    decoded[0] !== ED25519_MULTICODEC_PREFIX[0] ||
    decoded[1] !== ED25519_MULTICODEC_PREFIX[1]
  ) {
    throw new Error(
      "extractPublicKeyFromKeyDID: unexpected multicodec prefix",
    );
  }

  const rawKey = decoded.slice(ED25519_MULTICODEC_PREFIX.length);
  if (rawKey.length !== 32) {
    throw new Error(
      `extractPublicKeyFromKeyDID: expected 32 key bytes, got ${rawKey.length}`,
    );
  }
  return rawKey;
}

/**
 * Parses the DID method from a DID string.
 * Only "key" and "web" are supported.
 */
export function parseDIDMethod(did: string): DIDMethod {
  const parts = did.split(":");
  if (parts.length < 3 || parts[0] !== "did") {
    throw new Error(`parseDIDMethod: invalid DID format: ${did}`);
  }
  const method = parts[1];
  if (method !== "web" && method !== "key") {
    throw new Error(`parseDIDMethod: unsupported DID method "${method}"`);
  }
  return method;
}

/**
 * Converts a did:web DID to the canonical HTTPS URL for its DID document.
 *
 * did:web:example.com             => https://example.com/.well-known/did.json
 * did:web:example.com:agents:abc  => https://example.com/agents/abc/did.json
 */
export function webDIDToURL(did: string): string {
  const withoutScheme = did.slice("did:web:".length);
  if (!withoutScheme) {
    throw new Error(`webDIDToURL: empty did:web host in: ${did}`);
  }

  const colonIndex = withoutScheme.indexOf(":");
  if (colonIndex < 0) {
    const host = withoutScheme.replace(/%3A/gi, ":");
    return `https://${host}/.well-known/did.json`;
  }

  const host = withoutScheme.slice(0, colonIndex).replace(/%3A/gi, ":");
  const pathPart = withoutScheme.slice(colonIndex + 1).replaceAll(":", "/");
  return `https://${host}/${pathPart}/did.json`;
}

/**
 * Extracts the first Ed25519 public key from a DID document's verificationMethod array.
 * Returns the raw 32-byte key as a Uint8Array.
 */
export function extractPublicKeyFromDocument(doc: DIDDocument): Uint8Array {
  for (const vm of doc.verificationMethod) {
    if (vm.type !== "Ed25519VerificationKey2020") continue;
    if (!vm.publicKeyMultibase.startsWith("z")) {
      throw new Error(
        `extractPublicKeyFromDocument: expected base58btc multibase (prefix 'z'), got ${vm.publicKeyMultibase[0]}`,
      );
    }
    const decoded = decodeBase58Btc(vm.publicKeyMultibase.slice(1));
    if (decoded.length !== 32) {
      throw new Error(
        `extractPublicKeyFromDocument: unexpected key length ${decoded.length}`,
      );
    }
    return decoded;
  }
  throw new Error(
    `extractPublicKeyFromDocument: no Ed25519VerificationKey2020 in document for ${doc.id}`,
  );
}

/**
 * Validates that a parsed DIDDocument has the required structure.
 */
export function validateDIDDocument(
  raw: unknown,
  expectedDID: string,
): DIDDocument {
  if (typeof raw !== "object" || raw === null) {
    throw new Error("validateDIDDocument: response is not an object");
  }
  const doc = raw as Record<string, unknown>;

  if (doc["id"] !== expectedDID) {
    throw new Error(
      `validateDIDDocument: document id "${String(doc["id"])}" does not match requested DID "${expectedDID}"`,
    );
  }
  if (!Array.isArray(doc["verificationMethod"])) {
    throw new Error(
      "validateDIDDocument: verificationMethod is missing or not an array",
    );
  }

  return raw as DIDDocument;
}

File: python/src/aumos_identity/did.py
# SPDX-License-Identifier: BSL-1.1
# Copyright (c) 2026 MuVeraAI Corporation

"""DID document types, parsing, and derivation utilities."""

from __future__ import annotations

import base64
import re
from urllib.parse import urlparse

from .types import (
    DIDDocument,
    DIDMethod,
    UnsupportedDIDMethodError,
    VerificationMethod,
    VerificationMethodType,
)

# Multicodec prefix for Ed25519 public keys: 0xed 0x01
_ED25519_MULTICODEC_PREFIX = bytes([0xED, 0x01])

# Base58btc alphabet (Bitcoin alphabet)
_BASE58_ALPHABET = b"123456789ABCDEFGHJKLMNPQRSTUVWXYZabcdefghijkmnopqrstuvwxyz"


def encode_base58btc(data: bytes) -> str:
    """Encode bytes as base58btc (Bitcoin alphabet, no multibase prefix)."""
    count = 0
    for byte in data:
        if byte != 0:
            break
        count += 1

    n = int.from_bytes(data, "big")
    result = b""
    while n > 0:
        n, remainder = divmod(n, 58)
        result = bytes([_BASE58_ALPHABET[remainder]]) + result

    return (b"1" * count + result).decode("ascii")


def decode_base58btc(encoded: str) -> bytes:
    """Decode a base58btc string (no multibase prefix) to bytes."""
    n = 0
    for char in encoded:
        index = _BASE58_ALPHABET.find(char.encode("ascii"))
        if index < 0:
            raise ValueError(f"Invalid base58btc character: {char!r}")
        n = n * 58 + index

    leading_zeros = len(encoded) - len(encoded.lstrip("1"))
    raw = n.to_bytes((n.bit_length() + 7) // 8, "big") if n > 0 else b""
    return b"\x00" * leading_zeros + raw


def derive_key_did(public_key: bytes) -> str:
    """Derive a did:key DID from a raw 32-byte Ed25519 public key.

    Encoding: ``did:key:z`` + base58btc(0xed01 || public_key_bytes)
    """
    if len(public_key) != 32:
        raise ValueError(
            f"derive_key_did: expected 32-byte Ed25519 public key, got {len(public_key)}"
        )
    prefixed = _ED25519_MULTICODEC_PREFIX + public_key
    return "did:key:z" + encode_base58btc(prefixed)


def extract_public_key_from_key_did(did: str) -> bytes:
    """Extract the raw 32-byte Ed25519 public key from a did:key DID."""
    if not did.startswith("did:key:z"):
        raise ValueError(
            f"extract_public_key_from_key_did: not a base58btc did:key: {did}"
        )
    encoded = did[len("did:key:z"):]
    decoded = decode_base58btc(encoded)

    if len(decoded) < len(_ED25519_MULTICODEC_PREFIX):
        raise ValueError("extract_public_key_from_key_did: decoded bytes too short")

    prefix = decoded[: len(_ED25519_MULTICODEC_PREFIX)]
    if prefix != _ED25519_MULTICODEC_PREFIX:
        raise ValueError(
            f"extract_public_key_from_key_did: unexpected multicodec prefix {prefix.hex()}"
        )

    raw_key = decoded[len(_ED25519_MULTICODEC_PREFIX):]
    if len(raw_key) != 32:
        raise ValueError(
            f"extract_public_key_from_key_did: expected 32 key bytes, got {len(raw_key)}"
        )
    return raw_key


def parse_did_method(did: str) -> DIDMethod:
    """Extract and return the DID method. Only 'key' and 'web' are supported."""
    parts = did.split(":", 2)
    if len(parts) < 3 or parts[0] != "did":
        raise ValueError(f"parse_did_method: invalid DID: {did!r}")
    method = parts[1]
    if method == "key":
        return DIDMethod.KEY
    if method == "web":
        return DIDMethod.WEB
    raise UnsupportedDIDMethodError(f"unsupported DID method: {method!r}")


def web_did_to_url(did: str) -> str:
    """Convert a did:web DID to the canonical HTTPS URL for its DID document.

    Examples::

        did:web:example.com             -> https://example.com/.well-known/did.json
        did:web:example.com:agents:abc  -> https://example.com/agents/abc/did.json
    """
    without_scheme = did[len("did:web:"):]
    if not without_scheme:
        raise ValueError(f"web_did_to_url: empty did:web host in: {did!r}")

    colon_pos = without_scheme.find(":")
    if colon_pos < 0:
        host = without_scheme.replace("%3A", ":").replace("%3a", ":")
        return f"https://{host}/.well-known/did.json"

    host = without_scheme[:colon_pos].replace("%3A", ":").replace("%3a", ":")
    path_part = without_scheme[colon_pos + 1:].replace(":", "/")
    return f"https://{host}/{path_part}/did.json"


def extract_public_key_from_document(doc: DIDDocument) -> bytes:
    """Extract the first Ed25519 public key from a DID document.

    Returns the raw 32-byte key.
    """
    for vm in doc.verification_method:
        if vm.type != VerificationMethodType.ED25519_2020:
            continue
        if not vm.public_key_multibase.startswith("z"):
            raise ValueError(
                f"extract_public_key_from_document: expected base58btc multibase (prefix 'z'), "
                f"got {vm.public_key_multibase[:1]!r}"
            )
        decoded = decode_base58btc(vm.public_key_multibase[1:])
        if len(decoded) != 32:
            raise ValueError(
                f"extract_public_key_from_document: unexpected key length {len(decoded)}"
            )
        return decoded

    raise ValueError(
        f"extract_public_key_from_document: no Ed25519VerificationKey2020 "
        f"in document for {doc.id}"
    )


def build_key_did_document(did: str, public_key: bytes) -> DIDDocument:
    """Synthesize a DIDDocument for a did:key DID from its raw public key bytes.

    No network call is required. The document is constructed according to the
    did:key specification: one Ed25519VerificationKey2020 verification method,
    referenced in both ``authentication`` and ``assertionMethod``.

    Parameters
    ----------
    did:
        The fully-qualified ``did:key`` DID string.
    public_key:
        Raw 32-byte Ed25519 public key.

    Returns
    -------
    DIDDocument
    """
    if len(public_key) != 32:
        raise ValueError(
            f"build_key_did_document: expected 32-byte Ed25519 public key, "
            f"got {len(public_key)}"
        )
    # Multibase base58btc encoding: prefix 'z' + base58btc(0xed01 || public_key)
    prefixed = _ED25519_MULTICODEC_PREFIX + public_key
    public_key_multibase = "z" + encode_base58btc(prefixed)
    vm_id = f"{did}#key-1"

    verification_method = VerificationMethod(
        id=vm_id,
        type=VerificationMethodType.ED25519_2020,
        controller=did,
        public_key_multibase=public_key_multibase,
    )

    return DIDDocument(
        context=[
            "https://www.w3.org/ns/did/v1",
            "https://w3id.org/security/suites/ed25519-2020/v1",
        ],
        id=did,
        verification_method=[verification_method],
        authentication=[vm_id],
        assertion_method=[vm_id],
    )


def parse_did_document(raw: object, expected_did: str) -> DIDDocument:
    """Parse and validate a raw JSON-decoded object as a DIDDocument."""
    if not isinstance(raw, dict):
        raise ValueError(f"parse_did_document: expected dict, got {type(raw).__name__}")

    doc_id = raw.get("id")
    if doc_id != expected_did:
        raise ValueError(
            f"parse_did_document: document id {doc_id!r} does not match "
            f"requested DID {expected_did!r}"
        )

    raw_vms = raw.get("verificationMethod", [])
    if not isinstance(raw_vms, list):
        raise ValueError("parse_did_document: verificationMethod is not a list")

    verification_methods: list[VerificationMethod] = []
    for vm_raw in raw_vms:
        if not isinstance(vm_raw, dict):
            continue
        vm_type_str = vm_raw.get("type", "")
        try:
            vm_type = VerificationMethodType(vm_type_str)
        except ValueError:
            continue  # Skip unsupported verification method types.
        verification_methods.append(
            VerificationMethod(
                id=str(vm_raw.get("id", "")),
                type=vm_type,
                controller=str(vm_raw.get("controller", "")),
                public_key_multibase=str(vm_raw.get("publicKeyMultibase", "")),
            )
        )

    context_raw = raw.get("@context", [])
    context = list(context_raw) if isinstance(context_raw, list) else [str(context_raw)]

    auth_raw = raw.get("authentication", [])
    auth = list(auth_raw) if isinstance(auth_raw, list) else []

    assertion_raw = raw.get("assertionMethod", [])
    assertion = list(assertion_raw) if isinstance(assertion_raw, list) else []

    return DIDDocument(
        context=context,
        id=str(doc_id),
        verification_method=verification_methods,
        authentication=[str(a) for a in auth],
        assertion_method=[str(a) for a in assertion],
        created=raw.get("created"),
        updated=raw.get("updated"),
    )
